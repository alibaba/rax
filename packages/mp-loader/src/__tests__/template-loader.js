const { runLoaders } = require('loader-runner');

const templateLoaderPath = require.resolve('../template-loader');

describe('template-loader', () => {
  it('Template', (done) => {
    runLoaders({
      resource: __filename,
      loaders: [templateLoaderPath],
      readResource: (file, callback) => {
        callback(null, `
          <view>
            <text>hello world</text>
          </view>
        `);
      },
    }, (err, resource) => {
      expect(err).toBeNull();

      const { result, cacheable } = resource;
      expect(cacheable).toBeTruthy();
      expect(result).toMatchSnapshot();
      done();
    });
  });

  it('web-view', (done) => {
    runLoaders({
      resource: __filename,
      loaders: [templateLoaderPath],
      readResource: (file, callback) => {
        callback(null, `
          <web-view src="http://foo/bar" onMessage="handleMessage">
          </web-view>
        `);
      },
    }, (err, resource) => {
      expect(err).toBeNull();

      const { result, cacheable } = resource;
      expect(cacheable).toBeTruthy();
      expect(result).toMatchSnapshot();
      done();
    });
  });

  it('slot', (done) => {
    runLoaders({
      resource: __filename,
      loaders: [templateLoaderPath],
      readResource: (file, callback) => {
        callback(null, `
          <view slot="hello"></view>
        `);
      },
    }, (err, resource) => {
      expect(resource.result[0]).toMatch('return [_c(__components_ref__[\'view\']||\'view\',{slot:"hello"})];');
      done(err);
    });
  });

  it('options', (done) => {
    runLoaders({
      resource: 'template.axml',
      loaders: [templateLoaderPath + '?' + JSON.stringify({
        isEntryTemplate: true,
        cssPath: 'style.css',
        appCssPath: 'app.css',
      })],
      readResource: (file, callback) => {
        if (file.match(/\.css$/)) {
          callback(null, `
            .foo { color: red; }
          `);
        } else if (file.match(/\.axml$/)) {
          callback(null, `
            <view>
              <text>hello world</text>
            </view>
          `);
        }
      },
    }, (err, resource) => {
      expect(err).toBeNull();

      const { result, cacheable } = resource;
      expect(cacheable).toBeTruthy();
      expect(result).toMatchSnapshot();
      done();
    });
  });
});
