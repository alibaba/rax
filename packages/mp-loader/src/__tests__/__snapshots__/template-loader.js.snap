// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['template-loader Template 1'] = `
Array [
  "module.exports = function (__render__ ) {
    // Register render first
    var __render_helpers__ = (function() { var obj = require(\\"../../../render-helpers/src/index.js\\"); return obj && obj.__esModule ? obj.default : obj; })();
    __render_helpers__._r(__render__);
    var _c = __render_helpers__._c;var _s = __render_helpers__._s;var _l = __render_helpers__._l;var _t = __render_helpers__._t;var _m = __render_helpers__._m;var _v = __render_helpers__._v;var _e = __render_helpers__._e;var _cx = __render_helpers__._cx;var _u = __render_helpers__._u;;
  
    // Templates store
    var __templates_ref__ = {};
    function _w(is) { return __templates_ref__[is] ? __templates_ref__[is] : null; }
    
    // Custom components ref
    var __components_ref__ = {};
    
  
    
  
    function render(data) {
      
      return [_c(__components_ref__['view']||'view',[_c(__components_ref__['text']||'text',[_v(\\"hello world\\")])])];
    }
  
    return render;
  }
  
  
  ",
]
`;

exports['template-loader options 1'] = `
Array [
  "module.exports = function (__render__ ) {
    // Register render first
    var __render_helpers__ = (function() { var obj = require(\\"../packages/render-helpers/src/index.js\\"); return obj && obj.__esModule ? obj.default : obj; })();
    __render_helpers__._r(__render__);
    var _c = __render_helpers__._c;var _s = __render_helpers__._s;var _l = __render_helpers__._l;var _t = __render_helpers__._t;var _m = __render_helpers__._m;var _v = __render_helpers__._v;var _e = __render_helpers__._e;var _cx = __render_helpers__._cx;var _u = __render_helpers__._u;;
  
    // Templates store
    var __templates_ref__ = {};
    function _w(is) { return __templates_ref__[is] ? __templates_ref__[is] : null; }
    
    // Custom components ref
    var __components_ref__ = {};
    
  
    
  
    function render(data) {
      
      return _c('page', { 'data-userview-root': 'true' }, null, [_c(__components_ref__['view']||'view',[_c(__components_ref__['text']||'text',[_v(\\"hello world\\")])])]);
    }
  
    return render;
  }
  
  
  ",
]
`;

exports['template-loader web-view 1'] = `
Array [
  "module.exports = function (__render__ ) {
    // Register render first
    var __render_helpers__ = (function() { var obj = require(\\"../../../render-helpers/src/index.js\\"); return obj && obj.__esModule ? obj.default : obj; })();
    __render_helpers__._r(__render__);
    var _c = __render_helpers__._c;var _s = __render_helpers__._s;var _l = __render_helpers__._l;var _t = __render_helpers__._t;var _m = __render_helpers__._m;var _v = __render_helpers__._v;var _e = __render_helpers__._e;var _cx = __render_helpers__._cx;var _u = __render_helpers__._u;;
  
    // Templates store
    var __templates_ref__ = {};
    function _w(is) { return __templates_ref__[is] ? __templates_ref__[is] : null; }
    
    // Custom components ref
    var __components_ref__ = {};
    
  
    
  
    function render(data) {
      
      return [_c(__components_ref__['web-view']||'web-view',{onMessage:this.handleMessage})];
    }
  
    return render;
  }
  
  module.exports.getWebViewSource = function (data) { return \\"http://foo/bar\\"; };
  module.exports.getWebViewOnMessage = function (data) { return \\"handleMessage\\"; };
  ",
]
`;
